# Assignment 05
# Random search and HyperOpt

**Objectives:**

1. Load titanic.csv file.

2. Select ‘Pclass’,’SibSp’,’Parch’, ‘Fare’; columns as data. Select ‘Survived’ column as label.

3. Apply SVM to classify the data.

4. Use 5-Fold Cross Validation to generalise the model accuracy.

5. Use random search and hyperopt packages to identify the best possible SVM model from the following search space.

* 'C': [0.1, 1, 10, 100, 1000]
* 'gamma': [1, 0.1, 0.01, 0.001, 0.0001] 
* 'kernel': ['rbf']

In [1]:
# import libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.metrics import accuracy_score,classification_report,confusion_matrix

import warnings
warnings.filterwarnings('ignore')

## Loading Dataset 

In [2]:
df = pd.read_csv('titanic.csv')
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [4]:
#Select 'Pclass','SibSp','Parch','Fare' columns as data. Select ‘Survived’ column as label.

X = df[['Pclass','SibSp','Parch','Fare']]
y = df.Survived

In [5]:
X.head()

Pclass  SibSp  Parch     Fare
0       3      1      0   7.2500
1       1      1      0  71.2833
2       3      0      0   7.9250
3       1      1      0  53.1000
4       3      0      0   8.0500

**Inferences:**
* There are some null values in Age and Cabin columns.
* There are 891 observations and 12 features

## Model Building

In [6]:
# Apply SVM to classify the data

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20)

## SVM

In [7]:
from sklearn.svm import SVC
model_svc = SVC(kernel='linear', C=1, gamma='auto')
model_svc.fit(X_train,y_train)

y_pred = model_svc.predict(X_test)

print('Accuracy Score:\n',accuracy_score(y_test, y_pred))
print('Confusion Matrix:\n',confusion_matrix(y_test, y_pred))
print('Classification Report:\n',classification_report(y_test,y_pred))

Accuracy Score:
 0.6703910614525139
Confusion Matrix:
 [[97 22]
 [37 23]]
Classification Report:
               precision    recall  f1-score   support

           0       0.72      0.82      0.77       119
           1       0.51      0.38      0.44        60

    accuracy                           0.67       179
   macro avg       0.62      0.60      0.60       179
weighted avg       0.65      0.67      0.66       179



In [8]:
# Use 5-Fold Cross Validation to generalise the model accuracy.

from sklearn.model_selection import cross_val_score,KFold

kf=KFold(n_splits=5)
score=cross_val_score(model_svc,X,y,cv=kf)

print("Cross Validation Scores are: {}\n".format(score))

print("Average of Cross Validation Scores: {}".format(score.mean()))

Cross Validation Scores are: [0.61452514 0.65730337 0.66292135 0.7247191  0.74719101]

Average of Cross Validation Scores: 0.6813319942250958


In [9]:
# Use random search and hyperopt packages to identify the best possible SVM model from the following search space.
# 'C': [0.1, 1, 10, 100, 1000]
# 'gamma': [1, 0.1, 0.01, 0.001, 0.0001] 
# 'kernel': ['rbf']

search_space ={'C': [0.1, 1, 10, 100, 1000],
     'gamma': [1, 0.1, 0.01, 0.001, 0.0001],
     'kernel': ['rbf']}

## Random Search

In [10]:
from sklearn.model_selection import RandomizedSearchCV
random_search=RandomizedSearchCV(estimator = SVC(), param_distributions=search_space,verbose=2, n_jobs=-1,
                            n_iter=50)
random_result = random_search.fit(X_train, y_train)
print('Best Score: ', random_result.best_score_*100)
print('Best Params: ', random_result.best_params_)

Fitting 5 folds for each of 25 candidates, totalling 125 fits
Best Score:  71.36314389835518
Best Params:  {'kernel': 'rbf', 'gamma': 1, 'C': 1}


In [11]:
from sklearn.svm import SVC
model_svc = SVC(kernel='rbf', C=1000, gamma=0.001)
model_svc.fit(X_train,y_train)

y_pred = model_svc.predict(X_test)

print('Accuracy Score:\n',accuracy_score(y_test, y_pred))
print('Confusion Matrix:\n',confusion_matrix(y_test, y_pred))
print('Classification Report:\n',classification_report(y_test,y_pred))

Accuracy Score:
 0.6759776536312849
Confusion Matrix:
 [[91 28]
 [30 30]]
Classification Report:
               precision    recall  f1-score   support

           0       0.75      0.76      0.76       119
           1       0.52      0.50      0.51        60

    accuracy                           0.68       179
   macro avg       0.63      0.63      0.63       179
weighted avg       0.67      0.68      0.67       179



In [12]:
# Use 5-Fold Cross Validation to generalise the model accuracy.

from sklearn.model_selection import cross_val_score,KFold

kf=KFold(n_splits=5)
score=cross_val_score(model_svc,X,y,cv=kf)

print("Cross Validation Scores are: {}\n".format(score))

print("Average of Cross Validation Scores: {}".format(score.mean()))

Cross Validation Scores are: [0.63687151 0.64606742 0.73033708 0.71348315 0.74157303]

Average of Cross Validation Scores: 0.6936664365074383


## HypetOpt

In [13]:
from hyperopt import tpe, hp, fmin, STATUS_OK,Trials

# define objective function

def hyperparameter_tuning(params):
    clf = SVC(**params)
    acc = cross_val_score(clf, X, y,scoring="accuracy").mean()
    return {"loss": -acc, "status": STATUS_OK}


In [14]:
# Initialize trials object

search_space = {
    "C": hp.choice("C", [0.1, 1, 10, 100, 1000]),
    "gamma": hp.choice("gamma", [1, 0.1, 0.01, 0.001, 0.0001]),
    "kernel": hp.choice("kernel", ['rbf']),
}

trials = Trials()

best = fmin(
    fn=hyperparameter_tuning,
    space = search_space, 
    algo=tpe.suggest, 
    max_evals=50, 
    trials=trials
)

print("Best: {}".format(best))

100%|███████████████████████████████████████████████| 50/50 [02:06<00:00,  2.53s/trial, best loss: -0.7060134329295085]
Best: {'C': 4, 'gamma': 1, 'kernel': 0}


In [15]:
from sklearn.svm import SVC
model_svc = SVC(kernel='rbf', C=1000, gamma=0.1)
model_svc.fit(X_train,y_train)

y_pred = model_svc.predict(X_test)

print('Accuracy Score:\n',accuracy_score(y_test, y_pred))
print('Confusion Matrix:\n',confusion_matrix(y_test, y_pred))
print('Classification Report:\n',classification_report(y_test,y_pred))

Accuracy Score:
 0.6703910614525139
Confusion Matrix:
 [[93 26]
 [33 27]]
Classification Report:
               precision    recall  f1-score   support

           0       0.74      0.78      0.76       119
           1       0.51      0.45      0.48        60

    accuracy                           0.67       179
   macro avg       0.62      0.62      0.62       179
weighted avg       0.66      0.67      0.66       179



In [16]:
# Use 5-Fold Cross Validation to generalise the model accuracy.

from sklearn.model_selection import cross_val_score,KFold

kf=KFold(n_splits=5)
score=cross_val_score(model_svc,X,y,cv=kf)

print("Cross Validation Scores are: {}\n".format(score))

print("Average of Cross Validation Scores: {}".format(score.mean()))

Cross Validation Scores are: [0.67039106 0.63483146 0.75280899 0.70224719 0.7752809 ]

Average of Cross Validation Scores: 0.7071119201556713


**Findings:**

Cross Validation Score of **SVM** before hyperparameter tuning is : **0.6813**

|Tuning Method|'C'|Gamma|kernal|Cross Validation Score|
|-|-|-|-|-|
|Random Search|1000|rbf|0.001|0.6936|
HyperOpt|1000|rbf|0.1|**0.7071**|

##### Jibin K Joy, ML & AI, KKEM August 2022 Batch